optimum/habana/transformers/models/mixtral/modeling_mixtral.py

```diff
@@ -449,6 +449,192 @@ def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) ->
+            expert_routing_table=selected_experts,
+            router_weights=routing_weights,
+            w12=w12_list,
+            w3=w3_list,
+            permuted_weights=True,
+            activation="silu",
+            experts_min=0,
+            experts_max=7,
+        )
+        return final_hidden_states.view(-1, hidden_states.shape[1])
+
+
+def calculate_routing_tensors(
+    score: torch.Tensor, topk: int, hidden_states_dtype: torch.dtype
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    routing_weights = F.softmax(score, dim=1, dtype=torch.float32)
+    routing_weights, selected_experts = torch.topk(routing_weights, topk, dim=-1)
+    routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+    routing_weights = routing_weights.to(hidden_states_dtype)
+    return routing_weights, selected_experts
+
+
+class FusedMoE(nn.Module):
+    """FusedMoE layer for MoE models.
+    Based on implementation from vllm:
+    https://github.com/HabanaAI/vllm-fork/blob/84922944da5b65fa43bbdc465650911cbef9de72/vllm/hpu/ops.py#L282"""
+
+    def __init__(
+        self,
+        config: MixtralConfig,
+        params_dtype: Optional[torch.dtype] = None,
+    ):
+        super().__init__()
+        if params_dtype is None:
+            params_dtype = torch.get_default_dtype()
+
+        self.top_k = config.num_experts_per_tok
+        self.num_experts = config.num_local_experts
+        self.tp_size = 1
+        if is_deepspeed_available():
+            from deepspeed import comm as dist
+
+            if dist.is_initialized():
+                self.tp_size = dist.get_world_size()
+
+        self.intermediate_size = config.intermediate_size
+        self.intermediate_size_per_partition = config.intermediate_size // self.tp_size
+        self.hpu_static_fused_moe = DynamicFusedMOE(self.num_experts)
+
+        # Fused gate_up_proj (column parallel)
+        self.w13_weight = nn.Parameter(
+            torch.empty(
+                self.num_experts,
+                2 * self.intermediate_size_per_partition,
+                config.hidden_size,
+                dtype=params_dtype,
+                device="hpu" if self.tp_size > 1 else None,
+            ),
+            requires_grad=False,
+        )
+        self.register_parameter("w13_weight", self.w13_weight)
+        # down_proj (row parallel)
+        self.w2_weight = nn.Parameter(
+            torch.empty(
+                self.num_experts,
+                config.hidden_size,
+                self.intermediate_size_per_partition,
+                dtype=params_dtype,
+                device="hpu" if self.tp_size > 1 else None,
+            ),
+            requires_grad=False,
+        )
+        self.register_parameter("w2_weight", self.w2_weight)
+
+    def forward(self, hidden_states: torch.Tensor, router_logits: torch.Tensor) -> torch.Tensor:
+        final_hidden_states = self.hpu_static_fused_moe(
+            hidden_states, self.w13_weight, self.w2_weight, router_logits, self.top_k
```

Comment:
Please make sure that naming convention is not important here, i.e. it’s not based on some mathematical formula. You're calling the function with w13_weight and w2_weight, while in function definition you’re referring to these parameters as w12_weight and w3_weight. Could be a misspelling.